File: tools/feature-artifact-checker/statemodel_to_mermaid.py
# Does not travel with check_specs.py: this is a seed for the ADR-0035 rewrite,
# folded into the checker when it is built, adopted by nobody until then.
"""Deterministic spec State model (markdown table) -> Mermaid stateDiagram-v2.

Reads the `### State model` subsection ADR-0035 puts at the head of a spec's
functional requirements: the `*States:* … *Initial:* … *Terminal:* …`
declaration and the `| From | Trigger | Guard | To | FR ids |` table — or the
exact stateless declaration, which emits nothing.

Standard library only, no network, no wall clock, no environment reads: the
same input bytes produce the same output bytes on every run, which is what
makes a regenerate-and-diff gate over the emitted diagram trustworthy.
Validates the model-local ADR-0035 rules first and exits non-zero naming every
failure found, not only the first. The cross-checks that need the rest of the
spec — every WHILE names a declared state, every cited FR id exists and is an
event-driven, unwanted-behaviour or complex pattern — belong to the full
checker, not to this file.

Usage:
    python statemodel_to_mermaid.py <spec.md>       # mermaid to stdout
    python statemodel_to_mermaid.py --check-only <spec.md>
    python statemodel_to_mermaid.py --self          # embedded example + expected bytes
"""

import re
import sys

STATELESS = "This feature has no externally visible states."
FR_ID = re.compile(r"^FR-\d{3}$")


def parse(text):
    """Extract the state model from spec markdown. Returns (model, errors);
    model is None for the stateless declaration or on a parse failure."""
    section = text
    m = re.search(r"^###\s+State model\s*$(.*?)(?=^#|\Z)", text, re.M | re.S)
    if m:
        section = m.group(1)
    section = re.sub(r"<!--.*?-->", "", section, flags=re.S)  # comments never count

    if STATELESS in section:
        return None, []

    prose = " ".join(
        ln.strip() for ln in section.splitlines() if not ln.strip().startswith("|")
    )
    decl = re.search(
        r"\*States:\*\s*(?P<states>.*?)\s*\*Initial:\*\s*(?P<initial>.*?)\s*"
        r"\*Terminal:\*\s*(?P<terminal>.*?)\s*$",
        " ".join(prose.split()),
    )
    if not decl:
        return None, [
            "state model: neither a *States:*/*Initial:*/*Terminal:* declaration "
            f"nor the exact line {STATELESS!r} found"
        ]

    def name_list(raw):
        return [n.strip() for n in raw.rstrip(".").split(",") if n.strip()]

    rows, errors = [], []
    lines = [ln.strip() for ln in section.splitlines() if ln.strip().startswith("|")]
    for ln in lines[2:]:  # header and separator rows
        cells = [c.strip() for c in ln.strip("|").split("|")]
        if len(cells) != 5:
            errors.append(f"state model: table row has {len(cells)} cells, not 5: {ln}")
            continue
        frm, trigger, guard, to, frs = cells
        rows.append(
            {
                "from": frm,
                "trigger": trigger,
                "guard": "" if guard in ("—", "-", "") else guard,
                "to": to,
                "frs": [f.strip() for f in frs.split(",") if f.strip()],
            }
        )

    model = {
        "states": name_list(decl["states"]),
        "initial": name_list(decl["initial"])[0] if name_list(decl["initial"]) else None,
        "terminal": name_list(decl["terminal"]),
        "transitions": rows,
    }
    return model, errors


def validate(model):
    """Return a list of error strings; empty means the model is well-formed."""
    errors = []
    states = model["states"]
    if not states:
        return ["states: empty declaration"]
    seen = set()
    for name in states:
        if name in seen:
            errors.append(f"states: {name!r} declared more than once")
        seen.add(name)

    if model["initial"] not in seen:
        errors.append(f"initial: {model['initial']!r} is not a declared state")
    for name in model["terminal"]:
        if name not in seen:
            errors.append(f"terminal: {name!r} is not a declared state")

    keys = set()
    for i, t in enumerate(model["transitions"], 1):
        where = f"transition {i}"
        for endpoint in ("from", "to"):
            if t[endpoint] not in seen:
                errors.append(f"{where}: {endpoint}={t[endpoint]!r} is not a declared state")
        if not t["trigger"]:
            errors.append(f"{where}: empty trigger")
        if not t["frs"]:
            errors.append(f"{where}: cites no FR id")
        for fr in t["frs"]:
            if not FR_ID.match(fr):
                errors.append(f"{where}: {fr!r} is not an FR-nnn id")
        key = (t["from"], t["trigger"], t["guard"])
        if key in keys:
            errors.append(f"{where}: duplicate (from, trigger, guard) = {key!r} — ambiguous machine")
        keys.add(key)

    edges = {}
    for t in model["transitions"]:
        edges.setdefault(t["from"], []).append(t["to"])
    reachable, frontier = {model["initial"]}, [model["initial"]]
    while frontier:
        for nxt in edges.get(frontier.pop(0), []):
            if nxt not in reachable:
                reachable.add(nxt)
                frontier.append(nxt)
    for name in states:
        if name not in reachable:
            errors.append(f"reachability: {name!r} cannot be reached from the initial state")
        if name not in model["terminal"] and name not in edges:
            errors.append(f"dead end: non-terminal {name!r} has no outgoing transition")

    return errors


def mermaid_ids(states):
    """Deterministic state-name -> mermaid-identifier map; collisions get a
    numeric suffix in declaration order."""
    ids, taken = {}, set()
    for name in states:
        candidate = re.sub(r"[^A-Za-z0-9_]", "", name) or "S"
        base, n = candidate, 2
        while candidate in taken:
            candidate = f"{base}_{n}"
            n += 1
        ids[name], _ = candidate, taken.add(candidate)
    return ids


def emit(model):
    ids = mermaid_ids(model["states"])
    lines = ["stateDiagram-v2"]
    for name in model["states"]:  # declaration order — stable output
        if ids[name] != name:
            lines.append(f'    state "{name}" as {ids[name]}')
    lines.append(f"    [*] --> {ids[model['initial']]}")
    for t in model["transitions"]:  # table order — stable output
        label = t["trigger"]
        if t["guard"]:
            label += f" [{t['guard']}]"
        label += " / " + ",".join(t["frs"])
        lines.append(f"    {ids[t['from']]} --> {ids[t['to']]} : {label}")
    for name in model["terminal"]:
        lines.append(f"    {ids[name]} --> [*]")
    return "\n".join(lines) + "\n"


SELF_EXAMPLE = """### State model

*States:* Draft, Issued, Partially paid, Overdue, Paid, Cancelled.
*Initial:* Draft. *Terminal:* Paid, Cancelled.

| From | Trigger | Guard | To | FR ids |
|---|---|---|---|---|
| Draft | issue | customer account is active | Issued | FR-002 |
| Issued | payment received | covers open balance | Paid | FR-004 |
| Issued | payment received | below open balance | Partially paid | FR-005 |
| Partially paid | payment received | covers open balance | Paid | FR-005 |
| Issued | due date passed | — | Overdue | FR-007 |
| Partially paid | due date passed | — | Overdue | FR-007 |
| Overdue | payment received | covers open balance | Paid | FR-008 |
| Overdue | payment received | below open balance | Overdue | FR-008 |
| Issued | cancel | no payment recorded | Cancelled | FR-009 |
| Overdue | cancel | no payment recorded | Cancelled | FR-009 |
"""

SELF_EXPECTED = """stateDiagram-v2
    state "Partially paid" as Partiallypaid
    [*] --> Draft
    Draft --> Issued : issue [customer account is active] / FR-002
    Issued --> Paid : payment received [covers open balance] / FR-004
    Issued --> Partiallypaid : payment received [below open balance] / FR-005
    Partiallypaid --> Paid : payment received [covers open balance] / FR-005
    Issued --> Overdue : due date passed / FR-007
    Partiallypaid --> Overdue : due date passed / FR-007
    Overdue --> Paid : payment received [covers open balance] / FR-008
    Overdue --> Overdue : payment received [below open balance] / FR-008
    Issued --> Cancelled : cancel [no payment recorded] / FR-009
    Overdue --> Cancelled : cancel [no payment recorded] / FR-009
    Paid --> [*]
    Cancelled --> [*]
"""


def run_self():
    model, errors = parse(SELF_EXAMPLE)
    errors += validate(model) if model else []
    if errors:
        for e in errors:
            print(f"statemodel --self: {e}", file=sys.stderr)
        return 1
    got = emit(model)
    if got != SELF_EXPECTED:
        print("statemodel --self: emitted bytes differ from expected", file=sys.stderr)
        return 1
    print("statemodel --self: ok — parse, validate and emit reproduce the expected bytes")
    return 0


def main(argv):
    if "--self" in argv:
        return run_self()
    args = [a for a in argv[1:] if a != "--check-only"]
    if len(args) != 1:
        print(__doc__, file=sys.stderr)
        return 2
    with open(args[0], encoding="utf-8") as f:
        model, errors = parse(f.read())
    if model is None and not errors:
        return 0  # stateless declaration: nothing to emit, nothing to check here
    errors += validate(model) if model else []
    if errors:
        for e in errors:
            print(f"statemodel: {e}", file=sys.stderr)
        return 1
    if "--check-only" not in argv:
        sys.stdout.write(emit(model))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: tools/feature-artifact-checker/statemodel-to-mermaid.mjs
#!/usr/bin/env node
// Does not travel with check-specs.mjs: this is a seed for the ADR-0035 rewrite,
// folded into the checker when it is built, adopted by nobody until then.

const DOC = `Deterministic spec State model (markdown table) -> Mermaid stateDiagram-v2.

Reads the \`### State model\` subsection ADR-0035 puts at the head of a spec's
functional requirements: the \`*States:* … *Initial:* … *Terminal:* …\`
declaration and the \`| From | Trigger | Guard | To | FR ids |\` table — or the
exact stateless declaration, which emits nothing.

Node built-ins only, no network, no wall clock, no environment reads: the
same input bytes produce the same output bytes on every run, which is what
makes a regenerate-and-diff gate over the emitted diagram trustworthy.
Validates the model-local ADR-0035 rules first and exits non-zero naming every
failure found, not only the first. The cross-checks that need the rest of the
spec — every WHILE names a declared state, every cited FR id exists and is an
event-driven, unwanted-behaviour or complex pattern — belong to the full
checker, not to this file.

Usage:
    node statemodel-to-mermaid.mjs <spec.md>       # mermaid to stdout
    node statemodel-to-mermaid.mjs --check-only <spec.md>
    node statemodel-to-mermaid.mjs --self          # embedded example + expected bytes
`;

import { readFileSync } from "node:fs";
import process from "node:process";

const STATELESS = "This feature has no externally visible states.";
const FR_ID = /^FR-\d{3}$/;

// Python-repr-style quoting, for error messages that name a value exactly.
function repr(value) {
  if (value === null || value === undefined) return "None";
  return "'" + String(value).replaceAll("\\", "\\\\").replaceAll("'", "\\'") + "'";
}

// Extract the state model from spec markdown. Returns [model, errors];
// model is null for the stateless declaration or on a parse failure.
function parse(text) {
  let section = text;
  const m = /^###\s+State model\s*$([\s\S]*?)(?=^#|(?![\s\S]))/m.exec(text);
  if (m) {
    section = m[1];
  }
  section = section.replace(/<!--[\s\S]*?-->/g, ""); // comments never count

  if (section.includes(STATELESS)) {
    return [null, []];
  }

  const prose = section.split("\n")
    .filter((ln) => !ln.trim().startsWith("|"))
    .map((ln) => ln.trim())
    .join(" ");
  const decl =
    /\*States:\*\s*(?<states>.*?)\s*\*Initial:\*\s*(?<initial>.*?)\s*\*Terminal:\*\s*(?<terminal>.*?)\s*$/
      .exec(prose.split(/\s+/).filter(Boolean).join(" "));
  if (!decl) {
    return [null, [
      "state model: neither a *States:*/*Initial:*/*Terminal:* declaration " +
      `nor the exact line ${repr(STATELESS)} found`,
    ]];
  }

  const nameList = (raw) =>
    raw.replace(/\.+$/, "").split(",").map((n) => n.trim()).filter(Boolean);

  const rows = [];
  const errors = [];
  const lines = section.split("\n")
    .map((ln) => ln.trim())
    .filter((ln) => ln.startsWith("|"));
  for (const ln of lines.slice(2)) { // header and separator rows
    const cells = ln.replace(/^\|+/, "").replace(/\|+$/, "").split("|")
      .map((c) => c.trim());
    if (cells.length !== 5) {
      errors.push(`state model: table row has ${cells.length} cells, not 5: ${ln}`);
      continue;
    }
    const [frm, trigger, guard, to, frs] = cells;
    rows.push({
      from: frm,
      trigger,
      guard: ["—", "-", ""].includes(guard) ? "" : guard,
      to,
      frs: frs.split(",").map((f) => f.trim()).filter(Boolean),
    });
  }

  const initials = nameList(decl.groups.initial);
  const model = {
    states: nameList(decl.groups.states),
    initial: initials.length > 0 ? initials[0] : null,
    terminal: nameList(decl.groups.terminal),
    transitions: rows,
  };
  return [model, errors];
}

// Return a list of error strings; empty means the model is well-formed.
function validate(model) {
  const errors = [];
  const states = model.states;
  if (states.length === 0) {
    return ["states: empty declaration"];
  }
  const seen = new Set();
  for (const name of states) {
    if (seen.has(name)) {
      errors.push(`states: ${repr(name)} declared more than once`);
    }
    seen.add(name);
  }

  if (!seen.has(model.initial)) {
    errors.push(`initial: ${repr(model.initial)} is not a declared state`);
  }
  for (const name of model.terminal) {
    if (!seen.has(name)) {
      errors.push(`terminal: ${repr(name)} is not a declared state`);
    }
  }

  const keys = new Set();
  model.transitions.forEach((t, index) => {
    const where = `transition ${index + 1}`;
    for (const endpoint of ["from", "to"]) {
      if (!seen.has(t[endpoint])) {
        errors.push(`${where}: ${endpoint}=${repr(t[endpoint])} is not a declared state`);
      }
    }
    if (!t.trigger) {
      errors.push(`${where}: empty trigger`);
    }
    if (t.frs.length === 0) {
      errors.push(`${where}: cites no FR id`);
    }
    for (const fr of t.frs) {
      if (!FR_ID.test(fr)) {
        errors.push(`${where}: ${repr(fr)} is not an FR-nnn id`);
      }
    }
    const key = `(${[t.from, t.trigger, t.guard].map(repr).join(", ")})`;
    if (keys.has(key)) {
      errors.push(`${where}: duplicate (from, trigger, guard) = ${key} — ambiguous machine`);
    }
    keys.add(key);
  });

  const edges = new Map();
  for (const t of model.transitions) {
    if (!edges.has(t.from)) edges.set(t.from, []);
    edges.get(t.from).push(t.to);
  }
  const reachable = new Set([model.initial]);
  const frontier = [model.initial];
  while (frontier.length > 0) {
    for (const next of edges.get(frontier.shift()) ?? []) {
      if (!reachable.has(next)) {
        reachable.add(next);
        frontier.push(next);
      }
    }
  }
  for (const name of states) {
    if (!reachable.has(name)) {
      errors.push(`reachability: ${repr(name)} cannot be reached from the initial state`);
    }
    if (!model.terminal.includes(name) && !edges.has(name)) {
      errors.push(`dead end: non-terminal ${repr(name)} has no outgoing transition`);
    }
  }

  return errors;
}

// Deterministic state-name -> mermaid-identifier map; collisions get a
// numeric suffix in declaration order.
function mermaidIds(states) {
  const ids = new Map();
  const taken = new Set();
  for (const name of states) {
    let candidate = name.replace(/[^A-Za-z0-9_]/g, "") || "S";
    const base = candidate;
    let n = 2;
    while (taken.has(candidate)) {
      candidate = `${base}_${n}`;
      n += 1;
    }
    ids.set(name, candidate);
    taken.add(candidate);
  }
  return ids;
}

function emit(model) {
  const ids = mermaidIds(model.states);
  const lines = ["stateDiagram-v2"];
  for (const name of model.states) { // declaration order — stable output
    if (ids.get(name) !== name) {
      lines.push(`    state "${name}" as ${ids.get(name)}`);
    }
  }
  lines.push(`    [*] --> ${ids.get(model.initial)}`);
  for (const t of model.transitions) { // table order — stable output
    let label = t.trigger;
    if (t.guard) {
      label += ` [${t.guard}]`;
    }
    label += " / " + t.frs.join(",");
    lines.push(`    ${ids.get(t.from)} --> ${ids.get(t.to)} : ${label}`);
  }
  for (const name of model.terminal) {
    lines.push(`    ${ids.get(name)} --> [*]`);
  }
  return lines.join("\n") + "\n";
}

const SELF_EXAMPLE = `### State model

*States:* Draft, Issued, Partially paid, Overdue, Paid, Cancelled.
*Initial:* Draft. *Terminal:* Paid, Cancelled.

| From | Trigger | Guard | To | FR ids |
|---|---|---|---|---|
| Draft | issue | customer account is active | Issued | FR-002 |
| Issued | payment received | covers open balance | Paid | FR-004 |
| Issued | payment received | below open balance | Partially paid | FR-005 |
| Partially paid | payment received | covers open balance | Paid | FR-005 |
| Issued | due date passed | — | Overdue | FR-007 |
| Partially paid | due date passed | — | Overdue | FR-007 |
| Overdue | payment received | covers open balance | Paid | FR-008 |
| Overdue | payment received | below open balance | Overdue | FR-008 |
| Issued | cancel | no payment recorded | Cancelled | FR-009 |
| Overdue | cancel | no payment recorded | Cancelled | FR-009 |
`;

const SELF_EXPECTED = `stateDiagram-v2
    state "Partially paid" as Partiallypaid
    [*] --> Draft
    Draft --> Issued : issue [customer account is active] / FR-002
    Issued --> Paid : payment received [covers open balance] / FR-004
    Issued --> Partiallypaid : payment received [below open balance] / FR-005
    Partiallypaid --> Paid : payment received [covers open balance] / FR-005
    Issued --> Overdue : due date passed / FR-007
    Partiallypaid --> Overdue : due date passed / FR-007
    Overdue --> Paid : payment received [covers open balance] / FR-008
    Overdue --> Overdue : payment received [below open balance] / FR-008
    Issued --> Cancelled : cancel [no payment recorded] / FR-009
    Overdue --> Cancelled : cancel [no payment recorded] / FR-009
    Paid --> [*]
    Cancelled --> [*]
`;

function runSelf() {
  const [model, parseErrors] = parse(SELF_EXAMPLE);
  const errors = [...parseErrors, ...(model ? validate(model) : [])];
  if (errors.length > 0) {
    for (const e of errors) {
      console.error(`statemodel --self: ${e}`);
    }
    return 1;
  }
  const got = emit(model);
  if (got !== SELF_EXPECTED) {
    console.error("statemodel --self: emitted bytes differ from expected");
    return 1;
  }
  console.log("statemodel --self: ok — parse, validate and emit reproduce the expected bytes");
  return 0;
}

function main(argv) {
  if (argv.includes("--self")) {
    return runSelf();
  }
  const args = argv.filter((a) => a !== "--check-only");
  if (args.length !== 1) {
    console.error(DOC);
    return 2;
  }
  const [model, parseErrors] = parse(readFileSync(args[0]).toString("utf-8"));
  if (model === null && parseErrors.length === 0) {
    return 0; // stateless declaration: nothing to emit, nothing to check here
  }
  const errors = [...parseErrors, ...(model ? validate(model) : [])];
  if (errors.length > 0) {
    for (const e of errors) {
      console.error(`statemodel: ${e}`);
    }
    return 1;
  }
  if (!argv.includes("--check-only")) {
    process.stdout.write(emit(model));
  }
  return 0;
}

process.exit(main(process.argv.slice(2)));
